File: src/renderer/petRegistry.ts
export interface PetRegistryEntry {
  name: string;
  displayName: string;
  manifest: string;
}

interface PetListFile {
  default?: string;
  pets?: Array<{
    name?: string;
    displayName?: string;
    manifest?: string;
  }>;
}

const fallbackPets: PetRegistryEntry[] = [
  { name: "koda", displayName: "Koda", manifest: "/pets/koda/pet.json" },
];

export class PetRegistry {
  private readonly petsByName: Map<string, PetRegistryEntry>;
  readonly defaultPet: PetRegistryEntry;

  constructor(pets: PetRegistryEntry[], defaultName?: string) {
    const safePets = pets.length > 0 ? pets : fallbackPets;
    this.petsByName = new Map(safePets.map((pet) => [pet.name, pet]));
    this.defaultPet = this.petsByName.get(defaultName ?? "") ?? safePets[0];
  }

  get pets(): PetRegistryEntry[] {
    return Array.from(this.petsByName.values());
  }

  has(name: string): boolean {
    return this.petsByName.has(name);
  }

  get(name: string): PetRegistryEntry | undefined {
    return this.petsByName.get(name);
  }

  resolve(name: string | null | undefined): PetRegistryEntry {
    return name ? this.get(name) ?? this.defaultPet : this.defaultPet;
  }
}

export async function loadPetRegistry(): Promise<PetRegistry> {
  try {
    const response = await fetch("/pets/pets.json");

    if (!response.ok) {
      throw new Error(`pet list request failed: ${response.status} ${response.statusText}`);
    }

    const petList = normalizePetList((await response.json()) as PetListFile);
    return new PetRegistry(petList.pets, petList.defaultName);
  } catch (error) {
    console.error("[koda-desk] failed to load pet registry, using fallback", error);
    return new PetRegistry(fallbackPets, fallbackPets[0].name);
  }
}

function normalizePetList(value: PetListFile): { defaultName?: string; pets: PetRegistryEntry[] } {
  const pets = Array.isArray(value.pets)
    ? value.pets
        .filter((entry): entry is Required<Pick<PetRegistryEntry, "name" | "manifest">> & { displayName?: string } => {
          return typeof entry.name === "string" && entry.name.length > 0 && typeof entry.manifest === "string";
        })
        .map((entry) => ({
          name: entry.name,
          displayName: typeof entry.displayName === "string" && entry.displayName.length > 0
            ? entry.displayName
            : entry.name,
          manifest: entry.manifest,
        }))
    : [];

  return {
    defaultName: typeof value.default === "string" ? value.default : undefined,
    pets,
  };
}

File: src-tauri/src/pet_registry.rs
use std::fs;

use tauri::App;

const BUNDLED_PET_REGISTRY: &str = include_str!("../../public/pets/pets.json");

#[derive(Debug, Clone)]
pub struct PetRegistryEntry {
    pub name: String,
    pub display_name: String,
}

#[derive(Debug, serde::Deserialize)]
struct PetListFile {
    #[serde(default)]
    pets: Vec<PetListEntry>,
}

#[derive(Debug, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
struct PetListEntry {
    name: String,
    display_name: Option<String>,
}

pub fn load_pet_registry(app: &App) -> Vec<PetRegistryEntry> {
    read_pet_registry(app).unwrap_or_else(|error| {
        eprintln!("[koda-desk] failed to load pet registry, using fallback: {error}");
        fallback_pets()
    })
}

fn read_pet_registry(app: &App) -> Result<Vec<PetRegistryEntry>, String> {
    let content = read_pet_registry_content(app);
    let pet_list = serde_json::from_str::<PetListFile>(&content).map_err(|error| error.to_string())?;
    let pets = pet_list
        .pets
        .into_iter()
        .filter(|entry| !entry.name.is_empty())
        .map(|entry| PetRegistryEntry {
            display_name: entry.display_name.unwrap_or_else(|| entry.name.clone()),
            name: entry.name,
        })
        .collect::<Vec<_>>();

    if pets.is_empty() {
        return Err("pet registry does not contain any pets".to_string());
    }

    Ok(pets)
}

fn read_pet_registry_content(_app: &App) -> String {
    let source_path = std::env::current_dir()
        .ok()
        .map(|path| path.join("public").join("pets").join("pets.json"));

    if let Some(path) = source_path {
        if let Ok(content) = fs::read_to_string(path) {
            return content;
        }
    }

    BUNDLED_PET_REGISTRY.to_string()
}

fn fallback_pets() -> Vec<PetRegistryEntry> {
    vec![PetRegistryEntry {
        name: "koda".to_string(),
        display_name: "Koda".to_string(),
    }]
}
